doc/design/custom_xcode_schemes.md

```diff
@@ -0,0 +1,467 @@
+
+    Args:
+        targets: A `sequence` of target labels as `string` values.
+
+    Return:
+        A `struct` representing a build action.
+    """
+    pass
+
+def _test_action(targets):
+    """Constructs a test action for an Xcode scheme.
+
+    Args:
+        targets: A `sequence` of target labels as `string` values.
+
+    Return:
+        A `struct` representing a test action.
+    """
+    pass
+
+def _launch_action(target, args = None, env = None):
+    """Constructs a launch action for an Xcode scheme.
+
+    Args:
+        target: A target label as a `string` value.
+        args: Optional. A `list` of `string` arguments that should be passed to 
+            the target when executed.
+        env: Optional. A `dict` of `string` values that will be set as 
+            environment variables when the target is executed.
+
+    Return:
+        A `struct` representing a launch action.
+    """
+    pass
+
+xcode_schemes = struct(
+    scheme = _scheme,
+    build_action = _build_action,
+    test_action = _test_action,
+    launch_action = _launch_action,
+)
+
+```
+
+### Changes to `_xcodeproj` Rule
+
+#### `scheme_autogeneration_mode` Attribute
+
+The `scheme_autogeneration_mode` attribute determines how Xcode scheme autogeneration will occur.
+It's behavior is described in [Introduction of Scheme Autogeneration
+Mode](#introduction-of-scheme-autogeneration-mode).
+
+#### `schemes` Attribute
+
+The `schemes` attribute accepts a `sequence` of JSON `string` values as returned by
+`xcode_schemes.scheme`. The values are parsed and passed along to the scheme generation code in
+`rules_xcodeproj`. An Xcode scheme is generated for each entry in the list.
+
+### Changes to `xcodeproj` Macro 
+
+The `xcodeproj` macro will gain two parameters: `targets` and `schemes`.
+
+```python
+def xcodeproj(*, 
+    name, 
+    xcodeproj_rule = _xcodeproj, 
+    targets = [], 
+    schemes = None, 
+    **kwargs):
+    """Creates an .xcodeproj file in the workspace when run.
+
+    Args:
+        name: The name of the target.
+        xcodeproj_rule: The actual `xcodeproj` rule. This is overridden during
+            fixture testing. You shouldn't need to set it yourself.
+        targets: Optional. A `list` of targets to be included in the Xcode 
+            project.
+        schemes: Optional. A `list` of JSON `string` values as returned by
+            `xcode_schemes.scheme`.
+        **kwargs: Additional arguments to pass to `xcodeproj_rule`.
+    """
+    pass
+```
+
+In addition to what the `xcodeproj` macro does today, if a value is provided for the `schemes`
+parameter, the macro will collect all of the targets listed in the test and launch actions (not the
+build action) and add them to the overall targets list. We only collect targets from the test action
+and launch action because the target list for `xcodeproj` should only contain top-level or
+leaf-nodes (i.e., not depeneded upon by other targets in the Xcode project).
```

Comment:
> We only collect targets from the test action and launch action because the target list for `xcodeproj` should only contain top-level or leaf-nodes (i.e., not depeneded upon by other targets in the Xcode project).

A library _could_ be a top-level target though, and it might not have a test. I think it's fine to force them to re-list that target in `targets`, but it's probably worth mentioning.

Also, we need to make sure to allow duplications (i.e. someone listing out all top level targets in `targets` and specifying custom schemes for each).